<img src="https://upload.wikimedia.org/wikipedia/commons/4/47/Logo_UTFSM.png" width="200" alt="utfsm-logo" align="left"/>

# MAT281
### Aplicaciones de la Matemática en la Ingeniería

## Módulo 02
## Laboratorio Clase 04: Agrupando datos

### Instrucciones


* Completa tus datos personales (nombre y rol USM) en siguiente celda.
* La escala es de 0 a 4 considerando solo valores enteros.
* Debes _pushear_ tus cambios a tu repositorio personal del curso.
* Como respaldo, debes enviar un archivo .zip con el siguiente formato `mXX_cYY_lab_apellido_nombre.zip` a alonso.ogueda@gmail.com. 
* Se evaluará:
    - Soluciones
    - Código
    - Que Binder esté bien configurado.
    - Al presionar  `Kernel -> Restart Kernel and Run All Cells` deben ejecutarse todas las celdas sin error.
* __La entrega es al final de esta clase.__

__Nombre__: Gustavo Renato Arcaya Espinosa

__Rol__: 201610001-6

Se utilizará el mismo dataset de pokemon

In [16]:
import os
import pandas as pd

In [17]:
pkm = (
    pd.read_csv(os.path.join("data", "pokemon.csv"), index_col="#")
    .rename(columns=lambda x: x.replace(" ", "").replace(".", "_").lower())
)
pkm.head()

name  type1   type2  hp  attack  defense  sp_atk  sp_def  speed  \
#                                                                             
1      Bulbasaur  Grass  Poison  45      49       49      65      65     45   
2        Ivysaur  Grass  Poison  60      62       63      80      80     60   
3       Venusaur  Grass  Poison  80      82       83     100     100     80   
4  Mega Venusaur  Grass  Poison  80     100      123     122     120     80   
5     Charmander   Fire     NaN  39      52       43      60      50     65   

   generation  legendary  
#                         
1           1      False  
2           1      False  
3           1      False  
4           1      False  
5           1      False

## Ejercicio #1 (1 pto)

Agrupar por `generation` y `legendary` y obtener por grupo:

* Promedio de `hp`
* Mínimo y máximo de `sp_atk` y `sp_def`

In [18]:
(
    pkm.groupby('generation')
    .agg(
        {'legendary' : 'mean'}
    )
)

legendary
generation           
1            0.036145
2            0.047170
3            0.112500
4            0.107438
5            0.090909
6            0.097561

## Ejercicio #2 (1 pto)

El profesor Oakgueda determinó que una buen indicador de pokemones es: 

$$ 0.2 \, \textrm{hp} + 0.4 \,(\textrm{attack} + \textrm{sp_atk})^2 + 0.3 \,( \textrm{defense} + \textrm{sp_deff})^{1.5} + 0.1 \, \textrm{speed}$$

Según este indicador, ¿Qué grupo de pokemones (`type1`, `type2`) es en promedio mejor que el resto? 

In [19]:
pkm['IND_OAK'] = (0.2*pkm['hp']+
                            0.4 * (pkm['attack']+pkm['sp_atk'])**2+
                            0.3 * (pkm['defense'] + pkm['sp_def'])**(1.5) + 
                            0.1 * pkm['speed'])

In [20]:
pkm.groupby(['type1', 'type2']).agg({'IND_OAK': 'mean'}).sort_values('IND_OAK').tail(1)

IND_OAK
type1  type2              
Ground Fire   44774.854123

__Respuesta__: Ground	Fire

## Ejercicio #3 (1 pto)

Define una función que escale los datos tal que, si $s$ es una columna: 

$$s\_scaled = \frac{s - \min(s)}{\max(s) - \min(s)}$$

Y luego transforma cada columna agrupando por si el pokemon es legendario o no.

In [21]:
def minmax_scale(s):
    return (s-s.min())/(s.max()-s.min())

pkm.groupby('legendary').transform(minmax_scale)

hp    attack   defense    sp_atk    sp_def     speed  generation  \
#                                                                             
1    0.173228  0.244444  0.195556  0.333333  0.214286  0.258065         0.0   
2    0.232283  0.316667  0.257778  0.424242  0.285714  0.354839         0.0   
3    0.311024  0.427778  0.346667  0.545455  0.380952  0.483871         0.0   
4    0.311024  0.527778  0.524444  0.678788  0.476190  0.483871         0.0   
5    0.149606  0.261111  0.168889  0.303030  0.142857  0.387097         0.0   
6    0.224409  0.327778  0.235556  0.424242  0.214286  0.483871         0.0   
7    0.303150  0.438889  0.324444  0.600000  0.309524  0.612903         0.0   
8    0.303150  0.694444  0.471111  0.727273  0.309524  0.612903         0.0   
9    0.303150  0.550000  0.324444  0.903030  0.452381  0.612903         0.0   
10   0.169291  0.238889  0.266667  0.242424  0.209524  0.245161         0.0   
11   0.228346  0.322222  0.333333  0.333333  0.285714  0.341935         0.0   
12   0.307087  0.433333  0.422222  0.454545  0.404762  0.470968         0.0   
13   0.307087  0.544444  0.511111  0.757576  0.452381  0.470968         0.0   
14   0.173228  0.138889  0.133333  0.060606  0.000000  0.258065         0.0   
15   0.192913  0.083333  0.222222  0.090909  0.023810  0.161290         0.0   
16   0.232283  0.222222  0.200000  0.484848  0.285714  0.419355         0.0   
17   0.153543  0.166667  0.111111  0.060606  0.000000  0.290323         0.0   
18   0.173228  0.111111  0.200000  0.090909  0.023810  0.193548         0.0   
19   0.251969  0.472222  0.155556  0.212121  0.285714  0.451613         0.0   
20   0.251969  0.805556  0.155556  0.030303  0.285714  0.903226         0.0   
21   0.153543  0.222222  0.155556  0.151515  0.071429  0.329032         0.0   
22   0.244094  0.305556  0.222222  0.242424  0.142857  0.425806         0.0   
23   0.322835  0.416667  0.311111  0.363636  0.238095  0.619355         0.0   
24   0.322835  0.416667  0.333333  0.757576  0.285714  0.748387         0.0   
25   0.114173  0.283333  0.133333  0.090909  0.071429  0.432258         0.0   
26   0.212598  0.422222  0.244444  0.242424  0.238095  0.593548         0.0   
27   0.153543  0.305556  0.111111  0.127273  0.052381  0.419355         0.0   
28   0.251969  0.472222  0.266667  0.309091  0.195238  0.612903         0.0   
29   0.133858  0.305556  0.173333  0.181818  0.161905  0.322581         0.0   
30   0.232283  0.444444  0.284444  0.333333  0.280952  0.483871         0.0   
..        ...       ...       ...       ...       ...       ...         ...   
771  0.370079  0.333333  0.266667  0.606061  0.523810  0.354839         1.0   
772  0.303150  0.483333  0.311111  0.387879  0.204762  0.729032         1.0   
773  0.259843  0.294444  0.231111  0.430303  0.223810  0.619355         1.0   
774  0.192913  0.250000  0.644444  0.242424  0.619048  0.290323         1.0   
775  0.173228  0.250000  0.133333  0.272727  0.261905  0.225806         1.0   
776  0.263780  0.388889  0.213333  0.442424  0.442857  0.354839         1.0   
777  0.350394  0.527778  0.288889  0.606061  0.619048  0.483871         1.0   
778  0.220472  0.416667  0.382222  0.424242  0.319048  0.451613         1.0   
779  0.165354  0.361111  0.191111  0.242424  0.190476  0.212903         1.0   
780  0.330709  0.583333  0.315556  0.333333  0.295238  0.329032         1.0   
781  0.188976  0.338889  0.288889  0.206061  0.166667  0.296774         1.0   
782  0.169291  0.338889  0.288889  0.206061  0.166667  0.329032         1.0   
783  0.208661  0.338889  0.288889  0.206061  0.166667  0.264516         1.0   
784  0.228346  0.338889  0.288889  0.206061  0.166667  0.232258         1.0   
785  0.251969  0.472222  0.520000  0.290909  0.261905  0.509677         1.0   
786  0.212598  0.444444  0.520000  0.290909  0.261905  0.606452         1.0   
787  0.291339  0.500000  0.520000  0.290909  0.261905  0.412903         1.0   
788  0.330709  0.527778  0.520000  0.290909  0.261905  0.3161

## Ejercicio #4 (1 pto)

El profesor Oakgueda necesita saber cuántos pokemones hay luego de filtrar el dataset tal que el grupo de (`type1`, `type2`) tenga en promedio un indicador (el del ejercicio #2) mayor a 40000.

In [22]:
pkm.groupby(['type1', 'type2']).filter(lambda x : x['IND_OAK'].mean() > 40000 )

name    type1 type2   hp  attack  defense  sp_atk  sp_def  \
#                                                                          
425  Primal Groudon   Ground  Fire  100     180      160     150      90   
799   Hoopa Unbound  Psychic  Dark   80     160       60     170     130   

     speed  generation  legendary       IND_OAK  
#                                                
425     90           3       True  44774.854123  
799     80           6       True  44369.690779

__Respuesta:__ Hay solo dos pokemones que cumplen esta condición.

 * Primal Groudon	Ground	44774.854123
 * Hoopa Unbound	Psychic  44369.690779